### PR TITLE
#1602 Using "Hash-objectionaries" hash-map in Probe and Pull mojos

### PR DESCRIPTION
**Pull request by maxonfjvipon**
Ref: #1602 

What's done:
- Introduced "Hash-objectionary" hash-map that will be used in Probe in Pull mojos to probe and pull objects with versions
- Added function to get objectionary from the map by hash
- Repeated fields in Probe and Pull mojos are moved to the separated abstract class 
- Added one test for OyRemote to make sure that it can work with narrow hashes

---

### Detailed summary
- The PR introduces changes to the `OyRemote` class, `OyRemoteTest` class, `PullMojo` class, and `ProbeMojo` class.
- The `OyRemote` class no longer throws an `IOException` in its constructor.
- The `OyRemoteTest` class has a new test method `checksPresenceOfObjectWithNarrowHash()`.
- The `PullMojo` class has a new method `objectionaryByHash()` and a new field `objectionaries`.
- The `ProbeMojo` class has a new method `objectionaryByHash()` and a new field `objectionaries`.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by maxonfjvipon** [comment]
@volodya-lombrozo please review this one

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]
@maxonfjvipon Please, add human readable text which shortly describes what have been done in that PR. Since you have only the reference to the high-level issue #1602 - it isn't clear which part of this issue you are trying to solve.  The header `#1602 Objectionaries` doesn't say anything informative neither.  The same with `AI` generated description.

**Comment by maxonfjvipon** [comment]
@volodya-lombrozo done, please have a look

File: eo-maven-plugin/src/main/java/org/eolang/maven/ProbeMojo.java [hunk 1]
```diff
@@ -62,55 +63,17 @@
     defaultPhase = LifecyclePhase.PROCESS_SOURCES,
     threadSafe = true
 )
-public final class ProbeMojo extends SafeMojo {
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/main/java/org/eolang/maven/ProbeMojo.java:65]: @maxonfjvipon Maybe it's better to remove one of that mojos and make probing and pulling together in a single mojo? Actually we can use `composition` for it  (by creating two different objects and using it in a single mojo).
Inheritance it is suboptimal solution here, to be honest:
- https://www.yegor256.com/2016/09/13/inheritance-is-procedural.html
- https://en.wikipedia.org/wiki/Composition_over_inheritance

 Josh Bloch's book Effective Java 2nd Edition:
> Item 16: Favor composition over inheritance

File: eo-maven-plugin/src/main/java/org/eolang/maven/ProbeMojo.java [hunk 2]
```diff
@@ -152,6 +115,29 @@ public void exec() throws IOException {
         }
     }
 
+    /**
+     * Get objectionary by given hash from the map.
+     * @param hash Hash.
+     * @return Objectionary by given hash.
+     */
+    private Objectionary objectionaryBy(final CommitHash hash) {
+        return this.putIfAbsent(
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/main/java/org/eolang/maven/ProbeMojo.java:124]: @maxonfjvipon This code is misleading and it is the main drawback of the `inheritance`. Why `Mojo` has `putIfAbsent` method? Is `Mojo` a map? I don't think so.

File: eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java [hunk 3]
```diff
@@ -132,6 +91,38 @@ public void exec() throws IOException {
         );
     }
 
+    /**
+     * Get objectionary from the map by given hash.
+     * @param hash Hash.
+     * @return Objectionary by given hash.
+     */
+    private Objectionary objectionaryBy(final CommitHash hash) {
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:99]: @maxonfjvipon Maybe `objectionary` would be more suitable name here? 
https://www.yegor256.com/2018/08/22/builders-and-manipulators.html

File: eo-maven-plugin/src/main/java/org/eolang/maven/SafeMojoWithObjectionaries.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+     * @checkstyle VisibilityModifierCheck (5 lines)
+     */
+    @SuppressWarnings("PMD.ImmutableField")
+    protected Objectionary objectionary;
+
+    /**
+     * Hash-Objectionary map.
+     * @todo #1602:30min Use objectionaries to pull objects with different
+     *  versions. Objects with different versions are stored in different
+     *  storages (objectionaries). Every objectionary hash its own hash.
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/main/java/org/eolang/maven/SafeMojoWithObjectionaries.java:82]: @maxonfjvipon:
> Every objectionary hash its own hash. 

"has"?

File: eo-maven-plugin/src/main/java/org/eolang/maven/SafeMojoWithObjectionaries.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+     * @todo #1602:30min Use objectionaries to probe objects with different
+     *  versions. Objects with different versions are stored in different
+     *  storages (objectionaries). Every objectionary has its own hash.
+     *  To get versioned object from objectionary firstly we need to get
+     *  right objectionary by object's version and then get object from that
+     *  objectionary by name.
+     * @checkstyle MemberNameCheck (5 lines)
+     * @checkstyle VisibilityModifierCheck (4 lines)
+     */
+    protected final Map<String, Objectionary> objectionaries = new HashMap<>();
```
Comment (volodya-lombrozo) [inline eo-maven-plugin/src/main/java/org/eolang/maven/SafeMojoWithObjectionaries.java:95]: @maxonfjvipon Looks like we can have only one `hash` (a `CommitHash` instance), so I don't clearly understand why we need that map to save the only one entry?

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]

File: eo-maven-plugin/src/main/java/org/eolang/maven/SafeMojoWithObjectionaries.java (hunk 5 above)
Comment (maxonfjvipon) [reply eo-maven-plugin/src/main/java/org/eolang/maven/SafeMojoWithObjectionaries.java:95]: @volodya-lombrozo This PR is just preparation. Isn't it clear from `todo` what this map was added for?

File: eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java (hunk 3 above)
Comment (maxonfjvipon) [reply eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:99]: @volodya-lombrozo I get qulice warning: "Field objectionary has the same name as a method". I think `objectionaryBy` is a quite good name. It's still a noun, but with preposition, so we can read it as english sentence: objectionary by hash

**Comment by maxonfjvipon** [comment]
@volodya-lombrozo please have a look one more time

**Comment by volodya-lombrozo** [comment]
@Graur Could I ask you take a look, please?

File: eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java [hunk 6]
```diff
@@ -64,24 +66,17 @@ public final class PullMojo extends SafeMojo {
      * The Git hash to pull objects from, in objectionary.
      *
      * @since 0.21.0
+     * @checkstyle VisibilityModifierCheck (6 lines)
```
Comment (Graur) [inline eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:69]: @maxonfjvipon Could you please explain why we need to exclude the `VisibilityModifierCheck` here?

File: eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java [hunk 7]
```diff
@@ -64,24 +66,17 @@ public final class PullMojo extends SafeMojo {
...
-     * @since 0.10.0
-     */
-    @Parameter(property = "eo.overWrite", required = true, defaultValue = "false")
-    private boolean overWrite;
-
     /**
      * Read hashes from local file.
      *
      * @checkstyle MemberNameCheck (7 lines)
+     * @checkstyle VisibilityModifierCheck (6 lines)
```
Comment (Graur) [inline eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:79]: @maxonfjvipon Same here. Could you please explain why we need to exclude the `VisibilityModifierCheck` here?

File: eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java [hunk 8]
```diff
@@ -92,34 +87,49 @@ public final class PullMojo extends SafeMojo {
      * -DofflineHash=0.2.7:abc2sd3,0.2.8:s4se2fe
      *
      * @checkstyle MemberNameCheck (7 lines)
+     * @checkstyle VisibilityModifierCheck (6 lines)
```
Comment (Graur) [inline eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:90]: @maxonfjvipon Same here. Could you please explain why we need to exclude the VisibilityModifierCheck here?

File: eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java [hunk 9]
```diff
@@ -92,34 +87,49 @@ public final class PullMojo extends SafeMojo {
...
      */
     @Parameter(property = "offlineHash")
     private String offlineHash;
 
     /**
      * The objectionary.
+     *
+     * @checkstyle VisibilityModifierCheck (5 lines)
```
Comment (Graur) [inline eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:98]: @maxonfjvipon Same here. Could you please explain why we need to exclude the `VisibilityModifierCheck` here?

**Review by Graur — CHANGES_REQUESTED** [review]

File: eo-maven-plugin/src/main/java/org/eolang/maven/ProbeMojo.java [hunk 10]
```diff
@@ -152,6 +170,31 @@ public void exec() throws IOException {
         }
     }
 
+    /**
+     * Get objectionary by given hash from the map.
+     * @param hash Hash.
+     * @return Objectionary by given hash.
+     */
+    private Objectionary objectionaryBy(final CommitHash hash) {
```
[the remark this answers is not in the document]
Comment (maxonfjvipon) [reply eo-maven-plugin/src/main/java/org/eolang/maven/ProbeMojo.java:178]: @Graur please see [this comment](https://github.com/objectionary/eo/pull/2284#discussion_r1269703962)

File: eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java (hunk 6 above)
Comment (maxonfjvipon) [reply eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:69]: @Graur there's no need for it, thanks

**Comment by maxonfjvipon** [comment]
@Graur please have a look one more time

File: eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java (hunk 3 above)
Comment (Graur) [reply eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:145]: @maxonfjvipon Maybe it's better to make it complete then? For example: `objectionaryByHash()`. Like in `TranspileTojos.java` method `findByOptimized()`.
What if I call your method like this: `this.objectionaryBy(tag)`. Did it give me an objectionary by tag?

Comment (maxonfjvipon) [reply eo-maven-plugin/src/main/java/org/eolang/maven/PullMojo.java:145]: @Graur sounds better

**Review by Graur — APPROVED** [review]
@maxonfjvipon Looks good to me

**Comment by maxonfjvipon** [comment]
@yegor256 could you please have a look

**Comment by yegor256** [comment]
@rultor merge

**Pull request merged**